dm/relay/local_reader.go

```diff
@@ -567,66 +637,126 @@ func (r *BinlogReader) parseFile(
+		if err != nil {
+			return false, false, terror.Annotate(err, "decode relay meta toml file failed")
+		}
+		// current watched file size have no change means that no new writes have been made
+		// our relay meta file will be updated immediately after receive the rotate event,
+		// although we cannot ensure that the binlog filename in the meta is the next file after latestFile
+		// but if we return a different filename with latestFile, the outer logic (parseDirAsPossible)
+		// will find the right one
+		if meta.BinLogName != state.relayLogFile {
+			// we need check file size again, as the file may have been changed during our metafile check
+			cmp, err2 := fileSizeUpdated(state.fullPath, state.latestPos)
+			if err2 != nil {
+				return false, false, terror.Annotatef(err2, "latestFilePath=%s endOffset=%d", state.fullPath, state.latestPos)
+			}
+			switch {
+			case cmp < 0:
+				return false, false, terror.ErrRelayLogFileSizeSmaller.Generate(state.fullPath)
+			case cmp > 0:
+				return false, true, nil
+			default:
+				nextFilePath := filepath.Join(state.relayLogDir, meta.BinLogName)
+				log.L().Info("newer relay log file is already generated",
+					zap.String("now file path", state.fullPath),
+					zap.String("new file path", nextFilePath))
+				return false, false, nil
+			}
+		}
 
-	select {
-	case <-ctx.Done():
-		return false, false, 0, "", "", false, nil
-	case switchResp := <-switchCh:
-		// update new uuid
-		if err = r.updateUUIDs(); err != nil {
-			return false, false, 0, "", "", false, nil
+		// maybe UUID index file changed
+		switchPath, err := r.getSwitchPath()
+		if err != nil {
+			return false, false, err
 		}
-		return true, false, 0, switchResp.nextUUID, switchResp.nextBinlogName, false, nil
-	case updatePath := <-updatePathCh:
-		if strings.HasSuffix(updatePath, relayLogFile) {
-			// current relay log file updated, need to re-parse it
-			return false, true, latestPos, "", "", replaceWithHeartbeat, nil
+		if switchPath != nil {
+			// we need check file size again, as the file may have been changed during path check
+			cmp, err := fileSizeUpdated(state.fullPath, state.latestPos)
+			if err != nil {
+				return false, false, terror.Annotatef(err, "latestFilePath=%s endOffset=%d", state.fullPath, state.latestPos)
+			}
+			switch {
+			case cmp < 0:
+				return false, false, terror.ErrRelayLogFileSizeSmaller.Generate(state.fullPath)
+			case cmp > 0:
+				return false, true, nil
+			default:
+				log.L().Info("newer relay uuid path is already generated",
+					zap.String("current path", state.relayLogDir),
+					zap.Any("new path", switchPath))
+				return true, false, nil
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return false, false, nil
+		case <-r.Notified():
+			active, relayOffset = r.relay.IsActive(r.currentUUID, state.relayLogFile)
+			if active {
+				if relayOffset > state.latestPos {
+					return false, true, nil
+				}
+				// already read to relayOffset, try again
```

Comment:
can you explain the circumstances under which this would happen ?

seems  there is always new event(relayOffset > state.latestPos) when`r.Notified()`

Reply:
may already read, see 
https://github.com/D3Hunter/ticdc/blob/7bf4c84ab7380e8a09d606739b262d9e8827f082/dm/relay/local_reader.go#L830-L836

Reply:
another case: there's a gap between `file write` and `notify`, reader may read it in between them